File: go/misarmail/streaming.go
package misarmail

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// StreamEvent is one decoded Server-Sent Events frame.
//
// MisarMail emits unnamed frames — `data: {...}` with no `event:` line — so
// Event is normally empty. Data holds the decoded JSON, or Raw when the payload
// was not JSON.
type StreamEvent struct {
	// Event is the SSE `event:` name when the server sends one.
	Event string
	// Data is the decoded JSON payload. Nil when the payload was not JSON.
	Data map[string]any
	// Raw is the payload exactly as received, always populated.
	Raw string
}

// StreamingResource covers the two Server-Sent Events endpoints.
//
// Both live outside /v1 — /api/ai/generate-email/stream and
// /api/campaigns/{id}/send-stream — so they use the API base rather than the
// versioned one. Both are API-key authenticated and metered like any other
// call, so a plan refusal surfaces as *PlanLimitError here too.
type StreamingResource struct{ c *Client }

// GenerateEmail calls POST /api/ai/generate-email/stream and invokes onEvent
// for each frame until the stream ends.
//
// Returning an error from onEvent stops the stream and is returned to the
// caller — the way to break out early.
//
//	err := c.Streaming.GenerateEmail(ctx, map[string]any{"prompt": "…"},
//	    func(e misarmail.StreamEvent) error {
//	        fmt.Print(e.Data["delta"])
//	        return nil
//	    })
func (r *StreamingResource) GenerateEmail(ctx context.Context, body map[string]any, onEvent func(StreamEvent) error) error {
	return r.c.stream(ctx, http.MethodPost, "/ai/generate-email/stream", body, onEvent)
}

// CampaignSend calls GET /api/campaigns/{id}/send-stream, reporting progress
// while a campaign sends.
func (r *StreamingResource) CampaignSend(ctx context.Context, campaignID string, onEvent func(StreamEvent) error) error {
	path := "/campaigns/" + url.PathEscape(campaignID) + "/send-stream"
	return r.c.stream(ctx, http.MethodGet, path, nil, onEvent)
}

// stream opens an SSE connection and dispatches each frame to onEvent.
//
// Deliberately not retried: replaying a stream that failed mid-flight would
// duplicate whatever the caller already consumed.
func (c *Client) stream(ctx context.Context, method, path string, body map[string]any, onEvent func(StreamEvent) error) error {
	var bodyReader io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return err
		}
		bodyReader = bytes.NewReader(b)
	}

	req, err := http.NewRequestWithContext(ctx, method, c.apiBase+path, bodyReader)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("Accept", "text/event-stream")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return &NetworkError{Message: err.Error(), Cause: err}
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		raw, _ := io.ReadAll(resp.Body)
		if isPlanLimit(raw) {
			return planLimitFrom(resp, raw)
		}
		var envelope struct {
			Error string `json:"error"`
		}
		json.Unmarshal(raw, &envelope) //nolint:errcheck
		msg := envelope.Error
		if msg == "" {
			msg = strings.TrimSpace(string(raw))
		}
		if msg == "" {
			msg = resp.Status
		}
		return &APIError{Status: resp.StatusCode, Message: msg}
	}

	// A frame ends at a blank line, so scan line by line and flush on one.
	// bufio.Scanner's default 64KB token limit is ample for a frame; the buffer
	// is raised anyway because a single delta could in principle be large.
	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	var eventName string
	var dataLines []string

	flush := func() (bool, error) {
		if len(dataLines) == 0 {
			eventName = ""
			return false, nil
		}
		raw := strings.Join(dataLines, "\n")
		eventName, dataLines = "", nil

		if raw == "[DONE]" {
			return true, nil // sentinel: stop, do not dispatch
		}
		ev := StreamEvent{Event: eventName, Raw: raw}
		var decoded map[string]any
		if json.Unmarshal([]byte(raw), &decoded) == nil {
			ev.Data = decoded
		}
		return false, onEvent(ev)
	}

	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r")

		if line == "" {
			done, err := flush()
			if err != nil {
				return err
			}
			if done {
				return nil
			}
			continue
		}
		// Comments are keepalives.
		if strings.HasPrefix(line, ":") {
			continue
		}
		if rest, ok := strings.CutPrefix(line, "event:"); ok {
			eventName = strings.TrimSpace(rest)
			continue
		}
		if rest, ok := strings.CutPrefix(line, "data:"); ok {
			dataLines = append(dataLines, strings.TrimPrefix(rest, " "))
		}
	}

	if err := scanner.Err(); err != nil {
		return &NetworkError{Message: fmt.Sprintf("stream read: %v", err), Cause: err}
	}

	// A final frame with no trailing blank line.
	if _, err := flush(); err != nil {
		return err
	}
	return nil
}

File: typescript/src/resources/streaming.ts
import { MisarMailError } from "../errors.js";
import type { BaseClient } from "../types.js";

/**
 * One decoded Server-Sent Event frame.
 *
 * MisarMail's streams emit unnamed frames — `data: {...}` with no `event:`
 * line — and terminate with the sentinel `data: [DONE]`. The sentinel is
 * consumed by the iterator and never yielded, so callers can simply read to
 * completion.
 */
export interface MisarMailStreamEvent {
  /** The decoded JSON payload, or the raw string when it is not JSON. */
  data: unknown;
  /** The `event:` name when the server sends one. MisarMail currently does not. */
  event?: string;
}

/** A token of generated text, as emitted by the AI stream. */
export interface GenerateEmailDelta {
  delta?: string;
  type?: string;
  done?: boolean;
  error?: string;
}

export interface GenerateEmailOptions {
  prompt?: string;
  [key: string]: unknown;
}

/**
 * The streaming endpoints.
 *
 * Both live **off** the `/v1` prefix — they are `/api/ai/generate-email/stream`
 * and `/api/campaigns/{id}/send-stream` — so they go through the client's
 * root-base transport rather than the versioned one.
 *
 * Both are API-key authenticated (`verifyApiKey`) and metered like any other
 * call, so a plan refusal surfaces the same way it does elsewhere.
 */
export class StreamingResource {
  constructor(private readonly client: BaseClient) {}

  /**
   * `POST /api/ai/generate-email/stream` — token-by-token email generation.
   *
   * @example
   * for await (const chunk of blog.streaming.generateEmail({ prompt: "…" })) {
   *   if (chunk.delta) process.stdout.write(chunk.delta);
   * }
   */
  async *generateEmail(
    options: GenerateEmailOptions
  ): AsyncGenerator<GenerateEmailDelta, void, unknown> {
    for await (const frame of this.stream("POST", "/ai/generate-email/stream", options)) {
      yield frame.data as GenerateEmailDelta;
    }
  }

  /**
   * `GET /api/campaigns/{id}/send-stream` — live progress while a campaign sends.
   */
  async *campaignSend(
    campaignId: string
  ): AsyncGenerator<Record<string, unknown>, void, unknown> {
    const path = `/campaigns/${encodeURIComponent(campaignId)}/send-stream`;
    for await (const frame of this.stream("GET", path)) {
      yield frame.data as Record<string, unknown>;
    }
  }

  /**
   * Opens an SSE connection and yields each frame as it arrives.
   *
   * Frames are separated by a blank line and may span several `data:` lines,
   * which are joined with newlines per the SSE spec. A `[DONE]` payload ends
   * the stream.
   */
  async *stream(
    method: string,
    path: string,
    body?: unknown
  ): AsyncGenerator<MisarMailStreamEvent, void, unknown> {
    const res = await this.client.openStream(method, path, body);

    if (!res.ok || !res.body) {
      const raw = await res.text().catch(() => "");
      throw new MisarMailError(res.status, raw || res.statusText, "stream_error");
    }

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });

        // Frames end at a blank line. \r\n\r\n is tolerated for proxies that
        // rewrite line endings.
        let split: number;
        while ((split = findFrameEnd(buffer)) !== -1) {
          const frame = buffer.slice(0, split);
          buffer = buffer.slice(split).replace(/^(\r?\n){2}/, "");

          const parsed = parseFrame(frame);
          if (parsed === "done") return;
          if (parsed) yield parsed;
        }
      }

      // Whatever is left after the socket closes, if it is a complete frame.
      const tail = parseFrame(buffer);
      if (tail && tail !== "done") yield tail;
    } finally {
      reader.releaseLock();
    }
  }
}

function findFrameEnd(buffer: string): number {
  const lf = buffer.indexOf("\n\n");
  const crlf = buffer.indexOf("\r\n\r\n");
  if (lf === -1) return crlf;
  if (crlf === -1) return lf;
  return Math.min(lf, crlf);
}

/**
 * Parses one frame. Returns `"done"` for the terminating sentinel, `null` for
 * a comment-only or empty frame (SSE keepalives start with `:`).
 */
function parseFrame(frame: string): MisarMailStreamEvent | "done" | null {
  let event: string | undefined;
  const dataLines: string[] = [];

  for (const line of frame.split(/\r?\n/)) {
    if (!line || line.startsWith(":")) continue;
    if (line.startsWith("event:")) event = line.slice(6).trim();
    else if (line.startsWith("data:")) dataLines.push(line.slice(5).trimStart());
  }

  if (dataLines.length === 0) return null;

  const raw = dataLines.join("\n");
  if (raw === "[DONE]") return "done";

  try {
    return { data: JSON.parse(raw), ...(event ? { event } : {}) };
  } catch {
    return { data: raw, ...(event ? { event } : {}) };
  }
}
